<a href="https://colab.research.google.com/github/Bhagya-Navodyani/Intelligent_Fish_Storage-ML/blob/main/notebook/Fish_Storage_ML_Pipeline.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## 1. Dataset Loading

In [ ]:
import pandas as pd
import numpy as np

# CSV file already uploaded to Colab
filename = "/content/Final_csv.csv"

# Read CSV
df_raw = pd.read_csv(filename)

print("Uploaded file:", filename)
print("Rows, Columns:", df_raw.shape)

print("\nColumns:")
print(df_raw.columns.tolist())

df_raw.head()

Uploaded file: /content/Final_csv.csv
Rows, Columns: (10465, 20)

Columns:
['pc_timestamp', 'trial_id', 'fish_species', 'record_type', 'elapsed_sec', 'storage_temp_c', 'air_temp_c', 'humidity_pct', 'distance_cm', 'gas_raw', 'reed_raw', 'lid', 'current_lid_open_sec', 'total_lid_open_sec', 'lid_open_count', 'storage_time_hours', 'temp_difference_c', 'lid_open_ratio', 'lid_open_rate_per_hour', 'avg_lid_open_duration_sec']


,pc_timestamp,trial_id,fish_species,record_type,elapsed_sec,storage_temp_c,air_temp_c,humidity_pct,distance_cm,gas_raw,reed_raw,lid,current_lid_open_sec,total_lid_open_sec,lid_open_count,storage_time_hours,temp_difference_c,lid_open_ratio,lid_open_rate_per_hour,avg_lid_open_duration_sec
0,2026-09-29 10:02:23,TRIAL_20260929_100223,Linna,START,0,0.25,28.5,84.8,28.37,835,0,OPEN,0,0,1,0.0000,28.25,0.0,0.0,0.0
1,2026-09-29 10:02:53,TRIAL_20260929_100223,Linna,PERIODIC,30,0.25,28.3,84.1,28.38,835,0,OPEN,30,30,1,0.0083,28.05,1.0,0.0,30.0
2,2026-09-29 10:03:23,TRIAL_20260929_100223,Linna,PERIODIC,60,0.25,28.2,86.0,28.38,835,0,OPEN,60,60,1,0.0167,27.95,1.0,0.0,60.0
3,2026-09-29 10:03:53,TRIAL_20260929_100223,Linna,PERIODIC,90,0.25,28.1,86.0,28.37,835,0,OPEN,90,90,1,0.0250,27.85,1.0,0.0,90.0
4,2026-09-29 10:04:23,TRIAL_20260929_100223,Linna,PERIODIC,120,0.25,28.0,86.4,29.26,835,0,OPEN,120,120,1,0.0333,27.75,1.0,0.0,120.0


## 2. Dataset Audit

In [7]:
# Create a working copy while preserving the loaded dataset
df = df_raw.copy()

# Convert timestamp to datetime format
df["pc_timestamp"] = pd.to_datetime(
    df["pc_timestamp"],
    errors="coerce"
)

# Sort records chronologically
df = df.sort_values("pc_timestamp").reset_index(drop=True)

# Dataset summary
print("Total rows:", len(df))
print("Total columns:", len(df.columns))

print("\nMissing values in each column:")
print(df.isnull().sum())

print("\nFirst timestamp:", df["pc_timestamp"].min())
print("Last timestamp :", df["pc_timestamp"].max())

display(df.head())

Total rows: 10465
Total columns: 20

Missing values in each column:
pc_timestamp                   0
trial_id                       0
fish_species                   0
record_type                    0
elapsed_sec                    0
storage_temp_c                 0
air_temp_c                     0
humidity_pct                   0
distance_cm                  311
gas_raw                        0
reed_raw                       0
lid                            0
current_lid_open_sec           0
total_lid_open_sec             0
lid_open_count                 0
storage_time_hours             0
temp_difference_c              0
lid_open_ratio                 0
lid_open_rate_per_hour         0
avg_lid_open_duration_sec      0
dtype: int64

First timestamp: 2026-09-29 10:02:23
Last timestamp : 2026-10-03 00:22:24


,pc_timestamp,trial_id,fish_species,record_type,elapsed_sec,storage_temp_c,air_temp_c,humidity_pct,distance_cm,gas_raw,reed_raw,lid,current_lid_open_sec,total_lid_open_sec,lid_open_count,storage_time_hours,temp_difference_c,lid_open_ratio,lid_open_rate_per_hour,avg_lid_open_duration_sec
0,2026-09-29 10:02:23,TRIAL_20260929_100223,Linna,START,0,0.25,28.5,84.8,28.37,835,0,OPEN,0,0,1,0.0000,28.25,0.0,0.0,0.0
1,2026-09-29 10:02:53,TRIAL_20260929_100223,Linna,PERIODIC,30,0.25,28.3,84.1,28.38,835,0,OPEN,30,30,1,0.0083,28.05,1.0,0.0,30.0
2,2026-09-29 10:03:23,TRIAL_20260929_100223,Linna,PERIODIC,60,0.25,28.2,86.0,28.38,835,0,OPEN,60,60,1,0.0167,27.95,1.0,0.0,60.0
3,2026-09-29 10:03:53,TRIAL_20260929_100223,Linna,PERIODIC,90,0.25,28.1,86.0,28.37,835,0,OPEN,90,90,1,0.0250,27.85,1.0,0.0,90.0
4,2026-09-29 10:04:23,TRIAL_20260929_100223,Linna,PERIODIC,120,0.25,28.0,86.4,29.26,835,0,OPEN,120,120,1,0.0333,27.75,1.0,0.0,120.0


In [8]:
# Check exact duplicate rows
duplicate_count = df.duplicated().sum()

print("Exact duplicate rows:", duplicate_count)

# Display basic numerical statistics
display(df.describe())

Exact duplicate rows: 0


,pc_timestamp,elapsed_sec,storage_temp_c,air_temp_c,humidity_pct,distance_cm,gas_raw,reed_raw,current_lid_open_sec,total_lid_open_sec,lid_open_count,storage_time_hours,temp_difference_c,lid_open_ratio,lid_open_rate_per_hour,avg_lid_open_duration_sec
count,10465,10465.000000,10465.000000,10465.000000,10465.000000,10154.000000,10465.000000,10465.000000,10465.000000,10465.000000,10465.000000,10465.000000,10465.000000,10465.000000,10465.000000,10465.000000
mean,2026-10-01 05:07:33.561681664,155110.561682,1.828822,17.700554,81.426393,29.699733,473.731295,0.944386,19.286670,9724.143717,35.257047,43.086266,15.871732,0.080273,1.124134,265.053454
min,2026-09-29 10:02:23,0.000000,-0.250000,13.000000,57.400000,4.720000,200.000000,0.000000,0.000000,0.000000,1.000000,0.000000,1.900000,0.000000,0.000000,0.000000
25%,2026-09-30 07:29:29,77226.000000,0.000000,15.300000,73.300000,27.870000,275.000000,1.000000,0.000000,6551.000000,26.000000,21.451700,13.800000,0.054727,0.679600,266.136000
50%,2026-10-01 05:06:26,155043.000000,0.500000,17.300000,80.000000,28.710000,393.000000,1.000000,0.000000,10981.000000,37.000000,43.067500,15.650000,0.070716,0.837200,275.807000
75%,2026-10-02 02:48:26,233163.000000,1.500000,19.300000,89.200000,30.410000,627.000000,1.000000,0.000000,12762.000000,43.000000,64.767500,17.600000,0.080717,1.127900,296.784000
max,2026-10-03 00:22:24,310801.000000,23.500000,28.500000,100.000000,63.830000,1761.000000,1.000000,2278.000000,15721.000000,57.000000,86.333600,28.250000,1.000000,6.818200,630.000000
std,NaN,89930.812406,3.498060,3.010103,9.844427,3.760690,234.341593,0.229186,129.840186,4112.439112,12.845562,24.980780,3.476105,0.065577,0.872750,53.509667


### Initial Audit Findings

- The dataset contains 10,465 sensor observations.
- No exact duplicate rows were detected.
- The records are arranged chronologically using `pc_timestamp`.
- The `distance_cm` feature contains 311 missing values.
- The storage-temperature and gas-sensor columns require further anomaly investigation before model training.
- Sensor-quality issues will be handled during the preprocessing stage without modifying the original dataset.

In [9]:
# Check important sensor anomalies

zero_temp_count = (df["storage_temp_c"] == 0).sum()
high_gas_count = (df["gas_raw"] > 1000).sum()
missing_distance_count = df["distance_cm"].isnull().sum()

print("storage_temp_c = 0 rows:", zero_temp_count)
print("gas_raw > 1000 rows:", high_gas_count)
print("Missing distance_cm rows:", missing_distance_count)

storage_temp_c = 0 rows: 3128
gas_raw > 1000 rows: 14
Missing distance_cm rows: 311
